### ISSUE TITLE
JPEG: Unsupported JPEG process: SOF type 0xcf due to truncated stream

### ISSUE THREAD

**Issue by SamStephens** [haraldk/TwelveMonkeys#1054]
**Describe the bug**

When I try and load the [attached JPEG](https://github.com/user-attachments/assets/74a4f726-312e-47d6-8ff9-cfb7f1f80b3b), I get the the error Unsupported JPEG process: SOF type 0xcf.

As far as I can tell, this image is a valid JPEG. It opens in Gimp, a webtool I found claims it's valid, and I can open it with Pillow on Python.

**Version information**
1. The version of the TwelveMonkeys ImageIO library in use.
3.12.0

2. The *exact* output of `java --version` (or `java -version` for older Java releases). 

```
openjdk 21.0.5 2024-10-15 LTS
OpenJDK Runtime Environment Corretto-21.0.5.11.1 (build 21.0.5+11-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.5.11.1 (build 21.0.5+11-LTS, mixed mode, sharing)
```

3. Extra information about OS version, server version, standalone program or web application packaging, executable wrapper, etc. Please state exact version numbers where applicable.

My stack trace is from a unit test for a Gradle based project running on Ubuntu on WSL2.

**To Reproduce**

* Start a project including `com.twelvemonkeys.imageio:imageio-jpeg:3.12.0`.
* Have the image available in your path
* Simply load the image using `ImageIO.read`.

**Expected behavior**
The image to be loaded successfully.

**Example code**
Preferably as a failing JUnit test, or a standalone program with a `main` method that showcases the problem. 

```
    @Test
    public void canLoadJpegWithSofType0xcf() throws Exception {
        BufferedImage image = ImageIO.read(new File("/path/to/bad/image.jpg"));
    }
```

**Sample file(s)**

https://github.com/user-attachments/assets/74a4f726-312e-47d6-8ff9-cfb7f1f80b3b

**Stack trace**

```
javax.imageio.IIOException: Unsupported JPEG process: SOF type 0xcf
        at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readImage(Native Method)
        at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readInternal(JPEGImageReader.java:1378)
        at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:1162)
        at com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:355)
        at java.desktop/javax.imageio.ImageIO.read(ImageIO.java:1466)
        at java.desktop/javax.imageio.ImageIO.read(ImageIO.java:1363)
        at thumbnailfromimage.ThumbnailCreatorTest.x(ThumbnailCreatorTest.java:110)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
        at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
        at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
        at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:100)
        at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:530)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:758)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:453)
        at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:211)
```

**Screenshots**

N/A

**Additional context**

N/A

**Comment by SamStephens** [comment]
[This image](https://github.com/user-attachments/assets/fe288a6f-c918-4176-bd3a-82fc700554c9) has a similar issue but has SOF type 0xce

**Comment by haraldk** [comment]
Unfortunately, your sample files uses "arithmetic coding" (instead of the normal Huffman coding used by baseline JPEG). This part of the standard is not supported by the version of libjpeg that Java uses, due to patent issues. That means we can't read it either (I belive the patents are now expired, and the JDK could probably enable this feature, but it's not widely used so it probably won't happen). 

It might be possible to add support for this in the Java code that we use for lossless decoding, but I need to look deeper into that. Probably non-trivial. I might be able to help, but my time is limited these days. Feel free to look into adding this support and create a PR!

**Comment by haraldk** [comment]
PS: Upon further investigation, I'm no longer sure the images actually use arithmetic coding. It seems they are just standard Huffman ones with SOF0, but somehow the data is corrupted or misinterpreted... 🤔

**Comment by SamStephens** [comment]
@haraldk is there anything I can do to help you with this? Or should I leave it with you?

Also, I have images that appear okay but have the same issue but with other SOF types: "0xc5", "0xc6", "0xc8", "0xcb".

Would you like examples of these other images?

**Comment by haraldk** [comment]
I'm not sure what we can do about this... 

The problem is that the files are not valid JPEGs/complete JFIF files. They are somehow truncated, indicated by the missing EOI (end of image) marker. In addition, there seems to be some garbage data inserted at the end. I don't really know that this data is, but it contains byte sequences that is not allowed in the JPEG code stream (`0xFF` followed by something other than `0x00`, aka "byte stuffing"). 

However, almost everything needed to decode the images is there, so if you really want to, you can read the images like this:

```java
ImageReadParam param = reader.getDefaultReadParam();
int width = reader.getWidth(0);
int height = reader.getHeight(0);
BufferedImage image = reader.getRawImageType(0).createBufferedImage(width, height);
param.setDestination(image);

try {
    image = reader.read(0, param);
}
catch (IOException e) {
    // TODO: Optionally handle, log a warning/error etc
}
```

If you look at the image in other viewers/decoders, you will likely see that the lower right corner of the image is discolored. Reading with the above code will have the same problem.

It could be that images could be patched, by inserting an EOI marker at the right place, if you want to try that. Detecting this situation in code might be difficult though.

**Comment by SamStephens** [comment]
Thanks @haraldk .

I think there's one consideration before we can simply close this out. The core problem that lead to my reaching out is that I could not tell whether I was seeing exceptions because my files are malformed, or because of limitations in TwelveMonkeys ImageIO.

Do you think it is feasible and worthwhile to try and make these exceptions provide clarity as to whether they are caused by malformed files, or caused by wellformed files that use JPEG features TwelveMonkeys ImageIO does not support?

Alternatively, is there a tool or resource you could link to in your documentation that can definitively tell us whether an image is wellformed or not?

I did attempt to establish if these JPEGs were wellformed or not before creating this issue, and the several methods I tried made it look as though the JPEGs were wellformed, so my initial (incorrect) assumption filing the bug was that the issue was a TwelveMonkeys ImageIO limitation.

One more thing - you could consider talking about these issues in the Advanced Usage section of the README.md, including providing the code snippet for reading the image despite the exception being thrown.

**Comment by haraldk** [comment]
I agree that it would be better if the exception message was something like "Corrupt JPEG data: premature end of data segment" (which is given today as a warning only) or "Corrupt JPEG data: Missing EOI marker", rather than the current message about "Unsupported JPEG process"  that is just misleading. However, the way this is implemented, by delegating to the standard JPEG plugin makes this behavior hard to change.  The messages/exceptions are from the delegate, and probably from the native libjpeg code. 

I will give it some thought though... It is possible to scan for the EOI marker, but in 99% of the times this would be a waste of time, so I have ruled out this option. Obviously, it would be easier to control the exception messages if everything was done in Java and our own code, but writing a well performing JPEG decoder in Java is a big job...

Unfortunately, I don't know of an authoritative tool that can verify or validate a JPEG file. I use various tools depending on what I'm looking for. But often resorts to inspecting the files in a hex editor, like I did in this case. 

Good idea adding the code to the readme! Maybe you would like to create a PR? 😀 
In theory it should work for all plugins, but the result is very much plugin/implementation specific. I guess a common result would be an empty/blank image...
